src/model/UserphraseModel.cpp

```diff
@@ -218,3 +220,13 @@ const Userphrase *UserphraseModel::getUserphrase(const QModelIndex& idx)
 {
     return &userphrase_[idx.row()];
 }
+
+void UserphraseModel::undo()
+{
+    if(userphrase_tmp_.size() > 0) {
+        add(userphrase_tmp_[userphrase_tmp_.size()-1].phrase_, userphrase_tmp_[userphrase_tmp_.size()-1].bopomofo_);
```

Comment:
`userphrase_tmp_` can be simply implemented as a local variable somewhere in file `src/model/UserphraseModel.cpp`.